cellxgene_resources/curation_qa.ipynb

```diff
@@ -656,6 +656,133 @@
+    "library_id = 'batch'  # this variable will change depending on the dataset"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#Construct new expression dataframes for par and nonpar\n",
+    "for k,v in region_dict.items():\n",
+    "    a = v[0]\n",
+    "    df = pd.DataFrame()\n",
+    "    for gene in a.var.index:\n",
+    "        a.obs[gene] = a[:,a.var.index == gene].X[:,0].toarray()  # add each gene exp counts to obs\n",
+    "        df[gene] = a.obs.groupby(['donor_id', library_id])[gene].sum()  # sum counts of each gene by donor and library id\n",
+    "        v.append(df)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#Sum expression counts across regions\n",
+    "par_df = region_dict['par'][1]\n",
+    "par_df['par_sum'] = np.sum(par_df, axis=1)\n",
+    "nonpar_df = region_dict['nonpar'][1]\n",
+    "nonpar_df['nonpar_sum'] = np.sum(nonpar_df, axis=1)\n",
+    "nonpar_par = pd.merge(nonpar_df, par_df, how='left', left_index=True, right_index=True) \n",
+    "nonpar_par['nonPAR_to_PAR'] = nonpar_par['nonpar_sum']/nonpar_par['par_sum']  # calculate nonpar-to-par ratio"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#Generate donor sex summary dataframe\n",
+    "nonpar_par.reset_index(inplace=True)\n",
+    "counts_per_lib = nonpar_par.value_counts(['donor_id',library_id,'nonPAR_to_PAR']).reset_index().rename(columns={\"index\": \"value\", 0: \"count\"}) #this drops all na values\n",
+    "counts_per_lib.drop(columns=['count'],inplace=True)\n",
+    "counts_per_lib.set_index(['donor_id'], inplace=True)\n",
+    "counts_avg = counts_per_lib.groupby(level='donor_id').mean()  # collapse multi-index, calculating average expression counts across donor libraries/batches\n",
+    "counts_avg.reset_index(inplace=True)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#Assign sex using 0.5 threshold\n",
+    "def assign_sex(x):\n",
+    "    if x >= 0.5:\n",
+    "        return 'Male'\n",
+    "    elif x < 0.5:\n",
+    "        return 'Female'\n",
+    "    else:\n",
+    "        pass\n",
+    "    \n",
+    "counts_avg['scRNAseq_sex'] = counts_avg['nonPAR_to_PAR'].apply(assign_sex)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#Compare each donor's labeled sex and scRNAseq-determined sex\n",
+    "check = raw_adata.obs.value_counts(['donor_id', 'sex']).reset_index().rename(columns={\"index\": \"value\", 0: \"count\"})\n",
+    "check.drop(columns=['count'],inplace=True)\n",
+    "check_donors = pd.merge(counts_avg, check, how='left', left_on='donor_id', right_on='donor_id')\n",
+    "check_donors"
```

Comment:
I can envision a Dataset with many donors, rather than report the full donor eval, you'd instead only want to report the mismatches, or the ones that are currently "unknown". So the things to act on are less likely to get buried.